### PR TITLE
Fix consumer behaviour re: extend visibilty and delete

### PR DESCRIPTION
**Pull request by jharlap**
Previously the visibility timeout extender and message deleter were
middlewares that every user of sqsconsumer needed to wire in manually.
The timeout extender had a bug because it was middleware whereby if
previous messages took a long time to process, the visibility timeout
may have expired before the handler started executing, which includes
the visibility timeout extender.

PLAT-5420

### CODE REVIEW

File: cleanup_delete.go [hunk 1]
```diff
@@ -152,5 +124,29 @@ func (dq *deleteQueue) start(ctx context.Context) {
 				dq.deleteFromPending()
 			}
 		}
+		select {
+		case <-ctx.Done():
+			dq.drain()
+			// drain the delete queue and return
+			go func() {
+				for {
```
Comment (thebigjc) [inline cleanup_delete.go]: It isn't clear why this gofunc to grain is needed right after you call a message to drain the queue. Doesn't this gofunc leak as it never exits?

File: cleanup_visibility.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		case msg := <-veq.queue:
+			veq.removeFromPending(msg)
+		case <-veq.ticker:
+			veq.extendPending()
+		}
+
+		veq.Lock()
+		n := len(veq.entries)
+		veq.Unlock()
+		if n <= 0 {
```
Comment (thebigjc) [inline cleanup_visibility.go]: I'm pressure sure this can't be negative.

File: cleanup_visibility.go [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+	for i, e := range veq.entries {
+		if *msg.MessageId == *e.MessageId {
+			veq.entries = append(veq.entries[:i], veq.entries[i+1:]...)
+			return
+		}
+	}
+}
+
+// extendBatch extends the visibility of up to 10 messages and returns the list of messages that failed to extend and an error for overall failure.
```
Comment (thebigjc) [inline cleanup_visibility.go]: This function doesn't know anything about `10`

File: cleanup_delete.go (hunk 1 above)
Comment (jharlap) [inline cleanup_delete.go]: Yep, but as it's part of the shutdown process I don't care if it leaks 🙉

The drain func uses a timeout so there may be later sends to this channel that would block, which is what the leaked routine is for.

**Comment by jharlap** [comment]
The change is relatively complex because the program is continuously fetching more messages, but with back pressure if the workers are slow, and properly extending the visibility timeouts as well as deleting successfully processed messages. And it's doing a reasonable job of doing all these ops in batches to reduce SQS requests that cost money. It's probably easier to read the final state than to read the diff.

The purpose of the multiple contexts and wait groups is so that on shutdown all the in flight workers complete uninterrupted, and the deletes still happen. This minimizes the risk of unnecessary duplicate processing for messages during app deploys, or when a box is being replaced in an ASG.

I've run the tests with the race detector and I've also run some complex manual tests with sqsexec (the change to that program that uses this version of the lib is going to be another PR but I figured it'd be easier to do after this is merged)

**Comment by thebigjc** [comment]
Yeah, I get what it's doing - and I know we've talked about how this model sort of has to be complex  given how AWS and Go work. Overall probably needs some work to get it more readable, but happy to keep moving forward here.

**Comment by thebigjc** [comment]
🚢 🇮🇹

**Pull request merged**